### PR TITLE
[WASI] Remove unused fds before closing them.

### PR DESCRIPTION
**Pull request by MediosZ**
Fix PR #2662 so that the unused fds are closed successfully.

### CODE REVIEW

File: lib/host/wasi/wasifunc.cpp [hunk 1]
```diff
@@ -560,11 +560,11 @@ Expect<uint32_t> WasiFdAllocate::body(const Runtime::CallingFrame &, int32_t Fd,
 Expect<uint32_t> WasiFdClose::body(const Runtime::CallingFrame &, int32_t Fd) {
   const __wasi_fd_t WasiFd = Fd;
 
+  Env.close(WasiFd);
```
Comment (hydai) [inline lib/host/wasi/wasifunc.cpp:563]: Please add comments to describe why we have to `close` before `fdClose`.

**Review by hydai — CHANGES_REQUESTED** [review]

**Comment by MediosZ** [comment]
@hydai  
I agree that the `close` should be moved after `fdClose`. The problem is that after `fdClose`, the Node is removed from the FdMap, so that `close` cannot get the Fd to close.
Maybe we should add a function to get the Fd of the node, and then we will have something like this:
```c++
Expect<uint32_t> WasiFdClose::body(const Runtime::CallingFrame &, int32_t Fd) {
  const __wasi_fd_t WasiFd = Fd;

  // get the fd to be closed.
  __wasi_fd_t NodeFd;
  if (auto Res = Env.getNodeFd(WasiFd); unlikely(!Res)) {
    return Res.error();
  } else {
    NodeFd = *Res;
  }

  if (auto Res = Env.fdClose(WasiFd); unlikely(!Res)) {
    return Res.error();
  }
  // close the fd
  Env.close(NodeFd);

  return __WASI_ERRNO_SUCCESS;
}
``` 

I can confirm that this works, but I don't know if it's a good idea to expose the fd of the Node publicly.

**Comment by hydai** [comment]
@MediosZ 
How about moving the `Env.close` inside the `Env.fdClose` function if `Env.close` will always execute after `Env.fdClose`?

**Comment by MediosZ** [comment]
Agreed.
Seems that a separate function is unnecessary.

**Comment by hydai** [comment]
Hi @MediosZ 
Could you confirm again if the latest modification is still working in the examples?

**Comment by MediosZ** [comment]
@hydai I confirm the modifcation works.

**Comment by MediosZ** [comment]
This is what I get after running the `wasmedge_hyper_demo`
```
GET as byte stream: http://eu.httpbin.org/get?msg=Hello
Response: 200 OK
Headers: {
    "date": "Mon, 24 Jul 2023 02:05:51 GMT",
    "content-type": "application/json",
    "content-length": "237",
    "connection": "keep-alive",
    "server": "gunicorn/19.9.0",
    "access-control-allow-origin": "*",
    "access-control-allow-credentials": "true",
}

b"{\n  \"args\": {\n    \"msg\": \"Hello\"\n  }, \n  \"headers\": {\n    \"Host\": \"eu.httpbin.org\", \n    \"X-Amzn-Trace-Id\": \"Root=1-64bddc7f-19e3a62e0007c9342b8a2512\"\n  }, \n  \"origin\": \"121.249.15.204\", \n  \"url\": \"http://eu.httpbin.org/get?msg=Hello\"\n}\n"

GET and get result as string: http://eu.httpbin.org/get?msg=WasmEdge
{
  "args": {
    "msg": "WasmEdge"
  }, 
  "headers": {
    "Host": "eu.httpbin.org", 
    "X-Amzn-Trace-Id": "Root=1-64bddc80-51f9518828c785f15fef0085"
  }, 
  "origin": "121.249.15.204", 
  "url": "http://eu.httpbin.org/get?msg=WasmEdge"
}


POST and get result as string: http://eu.httpbin.org/post
with a POST body: hello wasmedge
{
  "args": {}, 
  "data": "hello wasmedge", 
  "files": {}, 
  "form": {}, 
  "headers": {
    "Content-Length": "14", 
    "Host": "eu.httpbin.org", 
    "X-Amzn-Trace-Id": "Root=1-64bddc81-6a44ff397d5fe7cf2a00fdc9"
  }, 
  "json": null, 
  "origin": "121.249.15.204", 
  "url": "http://eu.httpbin.org/post"
}

```

**Review by hydai — APPROVED** [review]

**Pull request merged**